holoviews/core/data/grid.py

```diff
@@ -301,6 +310,8 @@ def values(cls, dataset, dim, expanded=True, flat=True):
         if dim in dataset.vdims or dataset.data[dim.name].ndim > 1:
             data = dataset.data[dim.name]
             data = cls.canonicalize(dataset, data)
+            if da and isinstance(data, da.Array):
+                data = data.compute()
```

Comment:
When does this get called? On what data is this being called on?

Reply:
This should only be called during plotting, and only on the subset of data that you are actually displaying. If you use a datashader operation this won't end up being called at all, and all the processing should happen out-of-core.

Reply:
Ok, thanks for clarifying. Sounds reasonable.

Does this get cached?

Reply:
No, but if you've got ideas about caching I'd love to hear them. I'm currently running into issues with some remote datasets, which end up being downloaded multiple times.

Reply:
For the local scheduler there is http://dask.pydata.org/en/latest/caching.html

No opportunistic caching exists today for the distributed scheduler.  It's wouldn't be particularly hard to do though if someone wants to dive in.  Relevant issue here: https://github.com/dask/distributed/issues/681

Reply:
Naively was thinking we might be able to call `persist` here. It would still be blocking in some cases (single machine scheduler), but no worse than calling `compute`. Though given we would want the array as well (in cases where we would get Futures), we would need somewhere to hang on to the Futures. Perhaps in Holoviews (i.e. an object of some kind), a Distributed Client (if using Distributed), somewhere else?

Reply:
Is the use case here a multi-dimensional array which should be loaded into memory incrementally as you are exploring it? For example let's say you had a stack of images with x, y and z axes and you are browsing through the z-stack, you want each 2D array to persist in memory once it has been displayed? I think that would be a very useful option but I'd have to think about the best approach to support that. 

Currently our interfaces are stateless but I could imagine passing through a ``persist`` option to the interfaces in the constructor, which tells them to call ``.persist()`` before calling ``.compute()``. To make this concrete:

```python
darr = da.from_array(np.random.rand(30, 20, 20), (1, 1, 3))
ds = hv.Dataset((range(20), range(20), range(30), darr),
                ['x', 'y', 'z'], 'Value', persist=True)

z_stack = ds.to(hv.Image, ['x', 'y'])
z_stack[10]['Value']
```

In this example we create the 3D array, use ``.to`` to group it by the z-axis and then fetch the 2D array corresponding to z-value 10 and then use getitem to load the 'Value' array into memory. If we add the persist flag the array would then stay in memory and subsequent access to ``z_stack[10]['Value']`` would be much faster. Is this what you had in mind? If so we'll need to discuss whether interfaces could/should have state for backend specific options like this.

Reply:
I think that medium-term this is the sort of thing that should be handled on the Dask side.  We're well-positioned to solve this with minimal code-complexity elsewhere.  Again, you can probably do this on the local (non-distributed) scheduler today.

Reply:
It would be interesting to see how calling these lines ahead of time affected performance:

```python
from dask.cache import Cache
cache = Cache(2e9).register()
```

Reply:
To answer your question, @philippjfr, yes, basically as you described. We are looking at time instead of z (and the axis order differs), but for all intensive purposes this difference is merely semantic as far as Holoviews is concerned.

FWIW, as @mrocklin points out, we do have other options on the Dask side that could be helpful here or could be extended a bit (e.g. Cache support on Distributed). Issue ( https://github.com/dask/distributed/issues/681 ) outlines this problem a bit more. Alternatively there are some existing Dask options, which could allow this to be solved in user code.

There's a separate discussion that is a bit more interesting, which essentially involves having something like Holoviews drive computational priority in Dask. IOW one requests a large array be persisted (in memory, `persist`, or [on disk]( http://dask.pydata.org/en/latest/array-api.html#dask.array.store ), with `return_stored=True`), but would like to prioritize computations based on what a user would like to look at first. Outlined a bit more in issue ( https://github.com/dask/distributed/issues/1753 ).